clippy_lints/src/static_items_large_align.rs

```diff
@@ -0,0 +1,186 @@
+use rustc_span::Span;
+use rustc_typeck::hir_ty_to_ty;
+
+pub struct StaticItemsLargeAlign {
+    pub page_size: u64,
+}
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check for usage of static items which have type alignment larger than page size.
+    ///
+    /// ### Why is this bad?
+    /// Due to some known unsound issues, the type alignment may not be fulfilled.
+    /// For more information, see:
+    /// <https://github.com/rust-lang/rust/issues/70022> and
+    /// <https://github.com/rust-lang/rust/issues/70143>.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[repr(align(0x100000))]
+    /// struct Aligned(u8);
+    ///
+    /// static X: Aligned = Aligned(0);         // Bad
+    ///
+    /// fn main() {
+    ///     let x = Aligned(0);                 // Good
+    ///     println!("{:#x}", &x as *const _ as usize);
+    ///     println!("{:#x}", &X as *const _ as usize);
+    ///     let b = Box::new(Aligned(0));       // Good
+    ///     println!("{:#x}", Box::into_raw(b) as usize);
+    /// }
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub STATIC_ITEMS_LARGE_ALIGN,
+    pedantic,
+    "static items with large type alignment, which may trigger unsound problems"
+}
+impl_lint_pass!(StaticItemsLargeAlign => [STATIC_ITEMS_LARGE_ALIGN]);
+
+impl LateLintPass<'_> for StaticItemsLargeAlign {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if_chain! {
+            if let ItemKind::Static(hir_ty, _, _) = item.kind;
+            let ty = hir_ty_to_ty(cx.tcx, hir_ty);
+            let mut visited_tys = FxHashSet::default();
+            let mut intermediate_tys = Vec::new();
+            if let Some(la_ty) = self.check_ty_alignment(cx.tcx, ty, &mut visited_tys, &mut intermediate_tys);
+            then {
+                let mut span_notes: Vec<(Span, String)> = Vec::new();
+                if !intermediate_tys.is_empty() {
+                    let top_ty = intermediate_tys[0].ty;
+                    if !top_ty.is_adt() {
+                        span_notes.push((
+                            hir_ty.span,
+                            format!("this {} contains an inner type with large alignment", top_ty.prefix_string(cx.tcx)),
+                        ));
+                    }
+                    intermediate_tys.iter()
+                        .filter_map(|im_ty| Self::report_im_ty(cx, im_ty))
+                        .for_each(|ss| span_notes.push(ss));
+                }
+                span_notes.push(self.report_la_ty(cx, &la_ty));
+
+                span_lint_and_then(
+                    cx,
+                    STATIC_ITEMS_LARGE_ALIGN,
+                    item.span,
+                    "this static item (itself or its subfield) has a type alignment,\n\
+                    which is larger than page size and may not be fulfilled,\n\
+                    for more information, see <https://github.com/rust-lang/rust/issues/70022>.",
```

Comment:
```suggestion
                    "static item with large type alignment",
```

The lint message is not meant to explain the whole world :P Clippy will add a link to its documentation to the first lint emission anyway, so adding links is not necessary.